### PR TITLE
Fix close blocked by refcnt in test 0113.

### PR DESCRIPTION
**Pull request by emasab**
Breaks a circular dependency from
rko to rktp and back that prevents
the toppar being destroyed.

### CODE REVIEW

**Comment by milindl** [comment]
I see that the assertion fails, here https://confluentinc.semaphoreci.com/jobs/f3de9229-85de-4bb0-914f-3daf8b1f89d3#L5575

**Comment by emasab** [comment]
@milindl : seen that, I'm looking for a different fix.

File: src/rdkafka_cgrp.c [hunk 1]
```diff
@@ -2734,6 +2738,37 @@ static void rd_kafka_cgrp_partition_del(rd_kafka_cgrp_t *rkcg,
...
+                 * so rktp->rktp_fetchq->rkq_fwdq is NULL.
+                 * Purge remaining operations in rktp->rktp_fetchq->rkq_q,
+                 * while holding lock, to avoid circular references */
+                rkq = rktp->rktp_fetchq;
+                mtx_lock(&rkq->rkq_lock);
+                rd_assert(!rkq->rkq_fwdq);
+
+                next = TAILQ_FIRST(&rkq->rkq_q);
+                while ((rko = next)) {
+                        rd_assert(rko->rko_type == RD_KAFKA_OP_BARRIER);
```
Comment (milindl) [inline src/rdkafka_cgrp.c:2753]: Minor: Comment here explaining why we check only for barrier would be helpful

**Review by milindl — APPROVED** [review]
lgtm, with addition comment for more clarity requested

File: src/rdkafka_cgrp.c [hunk 2]
```diff
@@ -2734,6 +2738,37 @@ static void rd_kafka_cgrp_partition_del(rd_kafka_cgrp_t *rkcg,
...
+                        next = TAILQ_NEXT(next, rko_link);
+                        cnt++;
+                }
+
+                mtx_unlock(&rkq->rkq_lock);
+
+                if (cnt) {
+                        rd_kafka_dbg(rkcg->rkcg_rk, CGRP, "PARTDEL",
+                                     "Purge toppar buffer containing %d "
```
Comment (pranavrth) [inline src/rdkafka_cgrp.c:2762]: Lets add "fetch queue" as well in the debug logs.

File: src/rdkafka_cgrp.c [hunk 3]
```diff
@@ -2734,6 +2738,37 @@ static void rd_kafka_cgrp_partition_del(rd_kafka_cgrp_t *rkcg,
...
+                                     "barrier(s) to avoid circular references",
+                                     cnt);
+                        rd_kafka_q_purge(rktp->rktp_fetchq);
+                } else {
+                        rd_kafka_dbg(rkcg->rkcg_rk, CGRP, "PARTDEL",
+                                     "Not purging toppar buffer");
```
Comment (pranavrth) [inline src/rdkafka_cgrp.c:2768]: Lets add "fetch queue" as well in the debug logs.

**Review by pranavrth — CHANGES_REQUESTED** [review]

File: src/rdkafka_cgrp.c [hunk 4]
```diff
@@ -2734,6 +2738,45 @@ static void rd_kafka_cgrp_partition_del(rd_kafka_cgrp_t *rkcg,
...
+                /* Partition is being removed from the cluster and it's stopped,
+                 * so rktp->rktp_fetchq->rkq_fwdq is NULL.
+                 * Purge remaining operations in rktp->rktp_fetchq->rkq_q,
+                 * while holding lock, to avoid circular references */
+                rkq = rktp->rktp_fetchq;
+                mtx_lock(&rkq->rkq_lock);
+                rd_assert(!rkq->rkq_fwdq);
+
+                next = TAILQ_FIRST(&rkq->rkq_q);
+                while ((rko = next)) {
```
Comment (pranavrth) [inline src/rdkafka_cgrp.c:2752]: I don't think `next` is necessary. We can use:

```c
while (rko) {

   rko = TAILQ_NEXT(rko, rko_link);
}
```

File: src/rdkafka_cgrp.c [hunk 5]
```diff
@@ -2734,6 +2738,45 @@ static void rd_kafka_cgrp_partition_del(rd_kafka_cgrp_t *rkcg,
...
+                if (cnt) {
+                        rd_kafka_dbg(rkcg->rkcg_rk, CGRP, "PARTDEL",
+                                     "Purge toppar fetch queue buffer "
+                                     "containing %d message(s) to avoid "
+                                     "circular references",
+                                     cnt);
+                        rd_kafka_q_purge(rktp->rktp_fetchq);
+                } else {
+                        rd_kafka_dbg(rkcg->rkcg_rk, CGRP, "PARTDEL",
+                                     "Not purging toppar fetch queue buffer");
```
Comment (pranavrth) [inline src/rdkafka_cgrp.c:2776]: Can also add `No ops present in the buffer`.

**Review by pranavrth — COMMENTED** [review]
Looks good. Just some minor improvements.

**Comment by emasab** [comment]
@pranavrth no, it's just for leaving the CHANGELOG always ready for a release.

**Pull request merged**